custom/enikshay/tasks.py

```diff
@@ -0,0 +1,267 @@
+            try:
+                case_block = update.case_block()
+                if case_block:
+                    submit_case_blocks(
+                        [ElementTree.tostring(case_block.as_xml())],
+                        self.domain
+                    )
+            except Exception, e:
+                logger.error("Error calculating adherence values for episode case_id({}): {}".format(
+                    episode.case_id,
+                    e
+                ))
+
+    def _get_open_episode_cases(self):
+        # return all open 'episode' cases
+        case_accessor = CaseAccessors(self.domain)
+        case_ids = case_accessor.get_open_case_ids_in_domain_by_type(CASE_TYPE_EPISODE)
+        return case_accessor.iter_cases(case_ids)
+
+    @memoized
+    def get_doses_data(self):
+        # return 'doses_per_week' by 'schedule_id' from the Fixture data
+        fixtures = FixtureDataItem.get_indexed_items(self.domain, DAILY_SCHEDULE_FIXTURE_NAME, SCHEDULE_ID_FIXTURE)
+        return dict((k, int(fixture['doses_per_week'])) for k, fixture in fixtures.items())
+
+
+class EpisodeUpdate(object):
+    """
+    Class to capture adherence related calculations specific to an 'episode' case
+    """
+    def __init__(self, episode_case, case_updater):
+        """
+        Args:
+            episode_case: An 'episode' case object
+            case_repeater: EpisodeAdherenceUpdater object
+        """
+        self.episode = episode_case
+        self.case_updater = case_updater
+
+    def get_property(self, property):
+        """
+        Args:
+            name of the case-property
+
+        Returns:
+            value of the episode case-property named 'property'
+        """
+        return self.episode.dynamic_case_properties().get(property)
+
+    @memoized
+    def get_valid_adherence_cases(self):
+        # Returns list of 'adherence' cases of which 'adherence_value' is one of DOSE_KNOWN_INDICATORS
+        indexed_cases = get_reverse_indexed_cases_es(self.case_updater.domain, self.episode.case_id)
+        return [
+            case
+            for case in indexed_cases
+            if (not case.closed and case.type == CASE_TYPE_ADHERENCE and
+                case.dynamic_case_properties().get('adherence_value') in DOSE_KNOWN_INDICATORS)
+        ]
+
+    def get_latest_adherence_case_for_episode(self):
+        """
+        return open case of type 'adherence' reverse-indexed to episode that
+            has the latest 'adherence_date' property of all
+        """
+        # sometime far back
+        latest_date = pytz.UTC.localize(datetime.datetime(1990, 1, 1))
+        latest_case = None
+        for case in self.get_valid_adherence_cases():
+            adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date'))
+            if adherence_date > latest_date:
+                latest_date = adherence_date
+                latest_case = case
+        return latest_case
```

Comment:
Ha, now I see the issue you are pointing out. There is potential custom ES queries that can be written here. I will dig into this a bit more and come up with a proposal

@NoahCarnahan @snopoke